### PR TITLE
New extension WordHighlighter - for highlighting words.

### PR DESCRIPTION
**Pull request by lukasMega**
init commit for new extension WordHighlighter - for highlighting words.

discussion, how it started: https://github.com/FreshRSS/FreshRSS/discussions/6475

### CODE REVIEW

**Comment by lukasMega** [comment]
currently in this very first version (0.0.1) only one highlighting type (primary) is implemented:
Screen 2024-05-21 at 23 45 27

**Comment by lukasMega** [comment]
> Let me know how I can assist you.

@math-GH  I would appreciate if you go through PHP code because I'm not familiar with PHP ecosystem (I have used PHP probably 10 years ago). Thank you.

**Comment by lukasMega** [comment]
also I need help with fixing errors from PHPStan. And I don't know how to trigger automated checks via Github Actions / workflows.

**Comment by math-GH** [comment]
>  I don't know how to trigger automated checks via Github Actions / workflows.

First-time contributors needs a manual start from an repository admin. I do it for you

File: xExtension-WordHighlighter/extension.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		}
+
+		if (file_exists($configFileJson)) {
+			try {
+				$confJson = json_decode(file_get_contents($configFileJson) ?: '', true, 8, JSON_THROW_ON_ERROR);
+				$this->enable_in_article = (bool) $confJson['enable_in_article'] ?: false;
+				$this->enable_logs = (bool) $confJson['enable_logs'] ?: false;
+				$this->case_sensitive = (bool) $confJson['case_sensitive'] ?: false;
+				$this->separate_word_search = (bool) $confJson['separate_word_search'] ?: false;
+				$this->word_highlighter_conf = (string) implode("\n", (array) $confJson['words']);
```
Comment (math-GH) [inline xExtension-WordHighlighter/extension.php:85]: `Casting to string something that's already string. `

implode is already a string. A cast is not needed here

Comment (lukasMega) [reply xExtension-WordHighlighter/extension.php]: oh, I missed that. Thank you!

**Comment by Alkarex** [comment]
You can run `make test-all` locally.
See also https://github.com/FreshRSS/Extensions/blob/master/README.md#commands-for-developers

**Comment by math-GH** [comment]
The TR and FR translations does not to be supported if the translation is not done.
English/EN is fine for a start.

**Comment by lukasMega** [comment]
> You can run `make test-all` locally (this assumes that you have your `./Extensions/` and `./FreshRSS/` folders as siblings (i.e. next to each other in the same directory). See also https://github.com/FreshRSS/Extensions/blob/master/README.md#commands-for-developers

unfortunately I don't have proper DEV environment yet. I just edited files on my live running server (RPi5 via SSH) 

> The TR and FR translations does not to be supported if the translation is not done.
English/EN is fine for a start.

What are the process with translations? Should I delete non English translations? Or should I somehow (with help of Google Translate / DeepL) translate it to other supported languages?

File: xExtension-WordHighlighter/extension.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+			];
+			$configJson = json_encode($configObj, WordHighlighterExtension::JSON_ENCODE_CONF);
+			file_put_contents(join_path($staticPath, ('config.' . $current_user . '.json')), $configJson . PHP_EOL);
+			file_put_contents(join_path($staticPath, ('config.' . $current_user . '.js')), $this->jsonToJs($configJson) . PHP_EOL);
+		}
+
+		if (file_exists($configFileJson)) {
+			try {
+				$confJson = json_decode(file_get_contents($configFileJson) ?: '', true, 8, JSON_THROW_ON_ERROR);
+				if (json_last_error() !== JSON_ERROR_NONE || !is_array($confJson)) {
```
Comment (Alkarex) [inline xExtension-WordHighlighter/extension.php:81]: I do not believe `json_last_error() !== JSON_ERROR_NONE` will ever be relevant when you also use `JSON_THROW_ON_ERROR`. I believe the idea is to use one or the other, not both.

File: xExtension-WordHighlighter/extension.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		} elseif (Minz_Request::isPost()) {
+			$configWordList = html_entity_decode(Minz_Request::paramString('words_list'));
+
+			$this->word_highlighter_conf = $configWordList;
+			$this->enable_in_article = (bool) Minz_Request::paramString('enable-in-article');
+			$this->enable_logs = (bool) Minz_Request::paramString('enable_logs');
+			$this->case_sensitive = (bool) Minz_Request::paramString('case_sensitive');
+			$this->separate_word_search = (bool) Minz_Request::paramString('separate_word_search');
+
+			$lineSeparator = ((bool) strpos($configWordList, "\r\n")) ? "\r\n" : "\n";
```
Comment (lukasMega) [inline xExtension-WordHighlighter/extension.php:65]: I was surprised that on my linux machine (under debian and under docker) I got `\r\n` as line separator - maybe related to PHP somehow?

Comment (Frenzie) [reply xExtension-WordHighlighter/extension.php:65]: Got \r\n from where?

Comment (lukasMega) [reply xExtension-WordHighlighter/extension.php:65]: from variable `$configWordList` which is initialized by:
`$configWordList = html_entity_decode(Minz_Request::paramString('words_list'));` (few lines above)

Comment (Frenzie) [reply xExtension-WordHighlighter/extension.php:65]: And where did that come from? If it's a `<textarea>` for example those send \r\n.

Comment (lukasMega) [reply xExtension-WordHighlighter/extension.php:65]: you are right, it's from a `<textarea>` so I can remove that line and use directly `"\r\n"` right?

Comment (Frenzie) [reply xExtension-WordHighlighter/extension.php:65]: I'm not sure if all browsers are guaranteed to do that. Particularly I seem to recall Safari sending \n at some point. So I'd just keep it as is; doesn't hurt.

(Except I'd probably use something like `preg_split("/\r\n|\n|\r/", $etc)`. to keep it easier to read.)

Comment (lukasMega) [reply xExtension-WordHighlighter/extension.php]: I replaced it in my [last commit](https://github.com/FreshRSS/Extensions/pull/238/commits/a928549d50bbb77b0d6e8dcc51a6f61f3585d6d0) with regex based on your suggestion. Thank you @Frenzie.

**Comment by math-GH** [comment]
> > The TR and FR translations does not to be supported if the translation is not done.
> > English/EN is fine for a start.
> 
> What is the process with translations? Should I delete non English translations? Or should I somehow (with help of Google Translate / DeepL) translate it to other supported languages?

There is no process at all.

Minimum requirement is to have the EN (English) translation that is the fallback for all other languages.

If you like you could add `SK` (Slovakian) - as I can see in your profile it could be your native language.

If you would like to have German, French and Dutch as translation then ping @Alkarex , @Frenzie and me.

I would suggest not to use an automatic translation to keep the quality and context.

If someone wants to have his/her language in this extension then the source is open to send a PR ;)

**Comment by Frenzie** [comment]
In theory machine translation could be a decent starting point (marked as dirty of course), but in practice it'd probably be unnecessarily annoying to have to double check it against the original.

**Comment by Alkarex** [comment]
Sorry for the delay - I had forgotten about this PR...

**Comment by Alkarex** [comment]
https://github.com/FreshRSS/Extensions/pull/302

**Pull request merged**